jax_md/elasticity.py

```diff
@@ -0,0 +1,526 @@
+from typing import Dict, Callable, Union
+
+import jax
+import jax.numpy as jnp
+import jax.scipy as jsp
+from jax import jit, lax, grad, vmap, jacfwd, jacrev, jvp
+
+from jax_md import quantity
+from jax_md.util import Array
+from jax_md.util import f32
+from jax_md.util import f64
+
+
+def _get_strain_tensor_list(dim, dtype) -> Array:
+  if dim == 2:
+    strain_tensors = jnp.array([[[1, 0],[0, 0]],
+                                [[0, 0],[0, 1]],
+                                [[0, 1],[1, 0]],
+                                [[1, 0],[0, 1]],
+                                [[1, 1],[1, 0]],
+                                [[0, 1],[1, 1]]], dtype=dtype)
+  elif dim == 3:
+    strain_tensors = jnp.array([[[1, 0, 0], [0, 0, 0], [0, 0, 0]],
+                                [[0, 0, 0], [0, 1, 0], [0, 0, 0]],
+                                [[0, 0, 0], [0, 0, 0], [0, 0, 1]],
+                                [[0, 0, 0], [0, 0, 1], [0, 1, 0]],
+                                [[0, 0, 1], [0, 0, 0], [1, 0, 0]],
+                                [[0, 1, 0], [1, 0, 0], [0, 0, 0]],
+                                [[0, 0, 0], [0, 1, 0], [0, 0, 1]],
+                                [[1, 0, 0], [0, 0, 0], [0, 0, 1]],
+                                [[1, 0, 0], [0, 1, 0], [0, 0, 0]],
+                                [[1, 0, 0], [0, 0, 1], [0, 1, 0]],
+                                [[1, 0, 1], [0, 0, 0], [1, 0, 0]],
+                                [[1, 1, 0], [1, 0, 0], [0, 0, 0]],
+                                [[0, 0, 0], [0, 1, 1], [0, 1, 0]],
+                                [[0, 0, 1], [0, 1, 0], [1, 0, 0]],
+                                [[0, 1, 0], [1, 1, 0], [0, 0, 0]],
+                                [[0, 0, 0], [0, 0, 1], [0, 1, 1]],
+                                [[0, 0, 1], [0, 0, 0], [1, 0, 1]],
+                                [[0, 1, 0], [1, 0, 0], [0, 0, 1]],
+                                [[0, 0, 1], [0, 0, 1], [1, 1, 0]],
+                                [[0, 1, 0], [1, 0, 1], [0, 1, 0]],
+                                [[0, 1, 1], [1, 0, 0], [1, 0, 0]]], dtype=dtype)
+  else:
+    raise AssertionError('not implemented for {} dimensions'.format(dim))
+  return strain_tensors
+
+def _convert_responses_to_elastic_constants(response_all: Array) -> Array:
+  if response_all.shape[0] == 6:
+    cxxxx = response_all[0]
+    cyyyy = response_all[1]
+    cxyxy = 0.25 * response_all[2]
+    cxxyy = 0.5  * (response_all[3] - cxxxx - cyyyy)
+    cxxxy = 0.25 * (response_all[4] - 4 * cxyxy - cxxxx)
+    cyyxy = 0.25 * (response_all[5] - 4 * cxyxy - cyyyy)
+
+    C = jnp.array(
+        [[[[cxxxx, cxxxy], [cxxxy, cxxyy]],
+          [[cxxxy, cxyxy], [cxyxy, cyyxy]]],
+         [[[cxxxy, cxyxy], [cxyxy, cyyxy]],
+          [[cxxyy, cyyxy], [cyyxy, cyyyy]]]])
+
+  elif response_all.shape[0] == 21: 
+    cxxxx = response_all[0];
+    cyyyy = response_all[1];
+    czzzz = response_all[2];
+    cyzyz = response_all[3]/4.;
```

Comment:
Can you add spaces around binary operations (aside from power).

Reply:
done